Add std.internal.mmm and reap benefits in @safe-ty and OOM checking

JakobOvrum opened the pull request:
This patch adds a new internal module std.internal.mmm with the following functions:
```d
// Unsafe as T has pointers
T* checkedMalloc(T)() @system if (hasIndirections!T);

// Ditto
T[] checkedMallocArray(T)(size_t n) @system if (hasIndirections!T);

// Known to be safe because T has no pointers
T* checkedMalloc(T)() @trusted if (!hasIndirections!T);

// Ditto
T[] checkedMallocArray(T)(size_t n) @trusted if (!hasIndirections!T);

// Attributes depend on emplaceRef
T* checkedMalloc(T, Args...)(auto ref Args args);

// Attributes depend on the range primitives of `Range` as well as
// the construction and destruction of `ElementType!Range`
T[] checkedMallocArray(T, Range)(Range range)
if (isInputRange!Range && hasLength!Range && is(ElementType!Range : T));

// Ditto; same as above but with inferred element type
ElementType!Range[] checkedMallocArray(Range)(Range range)
if (isInputRange!Range && hasLength!Range);

// Always safe as calloc zero-initializes memory
T* checkedCalloc(T)() @trusted;

// Ditto
T[] checkedCallocArray(T)(size_t n) @trusted;

// Destroy all elements in the given array
void[] destroyArray(T)(ref T[] array);
```
Additionally, inferred purity can be supported if `malloc` and `calloc` are marked `pure` in core.stdc.stdlib in the spirit of memory allocation being practically pure.

Included are a few commits demonstrating how to reap benefits from this module. They can be moved to a different PR if desirable - the point is that they demonstrate the module's usefulness (hopefully).

It might be best to review this PR commit-wise.

edit:

These functions also handle OOM correctly; in Phobos, there are more examples of *incorrect* handling of OOM, like trying to *dynamically allocate* an exception, either directly or through `enforce`, rather than use `onOutOfMemoryError`.

JakobOvrum commented:
I'm aware of the test failures and I'll fix them as soon as I can.

JakobOvrum commented:
Test success depends on #3032.

WalterBright commented:
How does test coverage look?
```
dmd std/internal/mmm -unittest -main -cov
```

JakobOvrum commented:
> How does test coverage look?

I used DMD's coverage analysis when writing the tests. IIRC all paths are tested except the OOM paths. I'll confirm when all the dependencies are figured out.

JakobOvrum commented:
This passes now, finally. Review would be appreciated.

I want to patch `unique` and `RefCounted` to use this, but I don't know if I should put the commits in this PR or a new one that depends on this...

MartinNowak commented on std/algorithm/sorting.d:
```diff
@@ -1905,30 +1905,36 @@ schwartzSort(alias transform, alias less = "a < b",
+                // collectException(destroy(e)); // Issue 12647
+                try
+                    destroy(e);
+                catch(Exception) {}
```
You should have a utility function for this, destroyArrayReverse or so.

DmitryOlshansky replied:
+1

MartinNowak commented:
Looks great

DmitryOlshansky commented:
Would be awesome to address minor comments and move this in

schveiguy commented:
I have no skin in this game, but is there a better name we can use instead of `mmm`? That just doesn't seem right :)

JakobOvrum commented:
Updated:
 * tabs -> spaces
 * Added `destroyArray` utility function and changed `schwartzSort` to use it
 * Added some more tests
 * Added documentation (but the module is still internal)

Passes the autotester again now.

DmitryOlshansky commented:
Maybe we should make destroy array deallocate the array itself, keeping in mind the use cases in schwartzSort.

JakobOvrum commented:
> Maybe we should make destroy array deallocate the array itself, keeping in mind the use cases in schwartzSort.

Destruction and deallocation need to be separate so that the caller can manually vet the memory safety of the deallocation with `@trusted`, while leaving the memory safety of destruction to attribute inference.

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+
+Authors: Jakob Øvrum
+*/
+module std.internal.mmm;
```
I usually try to stay away from trivial matters, but `mmm` is quite out there. Where does it even come from?

JakobOvrum replied:
For posterity, it was meant to be "Manual Memory Management". It's an internal module so I didn't put much thought into it

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+
+import core.stdc.stdlib : calloc, malloc;
+import std.traits : hasElaborateDestructor, hasIndirections;
+import std.range.primitives;
```
list? (wouldn't ask if you didn't provide a list for the others)

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+
+    if (auto p = malloc(n))
+        return p[0 .. n];
+    else
```
superfluous `else`

JakobOvrum replied:
Well, it's a style choice. I think it's more structured than using early returns (although I like those when appropriate).

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+ *    n = number of objects to allocate space for
+ * Throws:
+ *   $(D OutOfMemoryError) on OOM
+ */
```
There's a third possibility here modeled after `minimallyInitializedArray`: allocate memory and zero only pointers but nothing else.

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+T* checkedMalloc(T)() @trusted
+    if (!hasIndirections!T)
+{
+    return cast(T*)checkedMallocArray(T.sizeof).ptr;
```
space before `checked`

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+}
+
+/**
+ * Construct a new T on the C heap with newly allocated space from $(D malloc).
```
"T" in code font

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+ * Throws:
+ *    $(D OutOfMemoryError) on OOM
+ */
+T* checkedMalloc(T, Args...)(auto ref Args args) // safety inferred
```
So this would be like `mallocConstruct` because it needs to be differentiated from the other one (e.g. if one wants to have a default-initialized type).

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+            {
+                // collectException(destroy(arr[i])); // Issue 12647
+                try
+                    destroy(arr[i]);
```
Something is off here - `arr[0]` will never be destroyed. Looks like a bug.

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+
+    if (auto p = calloc(T.sizeof, 1))
+        return cast(T*)p;
+    else
```
superfluous

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+void[] destroyArray(T)(ref T[] array) nothrow
+    if(hasElaborateDestructor!T)
+{
+    foreach_reverse(ref e; array)
```
space before `(`

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+
+    if (auto p = calloc(T.sizeof, n))
+        return (cast(T*)p)[0 .. n];
+    else
```
unneeded

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+ * Returns:
+ *    The _array as $(D void[]), useful for passing to a deallocation function
+ */
+void[] destroyArray(T)(ref T[] array) nothrow
```
not `nothrow`, D dtors may throw - leave it to inference

andralex commented on std/internal/mmm.d:
```diff
@@ -0,0 +1,337 @@
+    {
+        try
+            destroy(e);
+        catch(Exception) {}
```
oh sry - but why this? Information is lost here

JakobOvrum replied:
I'm not sure what the idea was. How should throwing from a destructor affect other destructors in the same scope?

andralex commented:
I think this needs to be redone as a layer working with `std.experimental.allocator`. Then the functionality in these functions is immediate to achieve by using `Mallocator`.

JakobOvrum commented:
If I'm not mistaken, the two interesting/useful functions in this are essentially the same as `make` and `makeArray` from std.alloactor, except hardcoded to use `malloc`. So we should definitely make this work with std.allocator instead.

JakobOvrum commented:
I see that `make` returns `null` on failure. In the example use cases from this PR, there is nothing that can be done except throw `OutOfMemoryError`, and you'd think this is fairly common in both libraries and applications when used with something like `malloc`. So maybe this use case should be reflected in std.allocator.package; a kind of `makeOrFail` wrapper function. It's orthogonal to memory safety but I mention it since it was part of this PR.
